packages/lesswrong/components/form-components/SocialPreviewUpload.tsx

```diff
@@ -0,0 +1,160 @@
+    backgroundColor: theme.palette.grey[100],
+    borderRadius: 6,
+    width: 'min-content',
+  },
+  title: {
+    fontSize: 14,
+    minHeight: 18,
+    fontWeight: 700,
+    marginTop: 12,
+    marginBottom: 7,
+  },
+  description: {
+    fontSize: 12,
+    minHeight: 36,
+    marginBottom: 7,
+  },
+  url: {
+    fontSize: 12,
+    color: theme.palette.grey[600],
+  },
+  blurb: {
+    paddingLeft: 16,
+    paddingRight: 16,
+    fontSize: 14,
+  }
+});
+
+// Derived from PostsEditQueryFragment, but with extra data that can be set locally (only dataWithDiscardedSuggestions)
+interface PostsEditWithLocalData extends PostsEdit { // fragment on Posts
+  readonly contents: RevisionEdit & { dataWithDiscardedSuggestions: string | null} | null,
+}
+
+/**
+ * Build the preview description and extract the first image in the document to use as a fallback. This is duplicating
+ * the fairly roundabout process that happens on the server to generate the preview description and image. The logic here
+ * is _almost_ exactly the same, but skips some overly complicated or potentially slow steps, namely:
+ * - trimming whitespace
+ * - rendering LaTex
+ * - handling markdown and DraftJS (seeing as these are no longer supported for non admins)
+ *
+ * The complete process:
+ * 1. The document is saved on the server (on submit or "save as draft")
+ *  1.1. originalContents is converted to html using dataToHtml, this handles LaTeX and trimming whitespace also (packages/lesswrong/server/editor/make_editable_callbacks.ts)
+ *  1.2. socialPreviewImageAutoUrl is set to the first image in the document (packages/lesswrong/server/callbacks/postCallbacks.ts)
+ * 2. The document is loaded
+ *  2.1. plaintextDescription is resolved from the html (packages/lesswrong/server/resolvers/revisionResolvers.ts)
+ *  2.2. socialPreviewImageUrl is resolved from the combination of socialPreviewImageId (the image that is (maybe) set explicitly) and socialPreviewAutoUrl (packages/lesswrong/lib/collections/posts/schema.tsx)
+ * 3. The head tags are set in PostsPage (packages/lesswrong/components/posts/PostsPage/PostsPage.tsx)
+ *  3.1 plaintextDescription is truncated
+ *  3.2 socialPreviewImageUrl is just used directly
+ */
+const buildPreviewFromDocument = (document: PostsEditWithLocalData): {description: string | null, fallbackImageUrl: string | null} => {
+  const dataWithDiscardedSuggestions = document.contents?.dataWithDiscardedSuggestions
+  const originalContents = document.contents?.originalContents
+
+  if (!originalContents) return {description: null, fallbackImageUrl: null}
+  if (originalContents.type !== "html" && originalContents.type !== "ckEditorMarkup") {
+    return {
+      description: `<Description preview not supported for this editor type (${originalContents.type}), switch to HTML or EA Forum Docs [Beta] to see the description preview>`,
+      fallbackImageUrl: null,
+    };
+  }
+
+  const html = dataWithDiscardedSuggestions ?? originalContents.data
```

Comment:
Should we be defaulting to `dataWithDiscardedSuggestions`? I'm kind of unsure but I guess that's the right thing to do here, to ignore suggestions when building the preview data. (Unrelated to this PR, but I think this is a really bad variable name, since I keep thinking that it's the opposite of what it's meant to be, but I guess it's just to [match CKEditor's naming](https://ckeditor.com/docs/ckeditor5/latest/api/module_track-changes_trackchangesdata-TrackChangesData.html#function-getDataWithDiscardedSuggestions))

Reply:
This is what we use on the backend to create the preview description so I think it makes sense to do so here